api/envoy/config/filter/network/tcp_proxy/v2/tcp_proxy.proto

```diff
@@ -21,19 +22,65 @@ message TcpProxy {
   // <config_network_filters_tcp_proxy_stats>`.
   string stat_prefix = 1 [(validate.rules).string.min_bytes = 1];
 
-  // The upstream cluster to connect to.
-  //
-  // .. note::
-  //
-  //  Once full filter chain matching is implemented in listeners, this field will become the only
-  //  way to configure the target cluster. All other matching will be done via :ref:`filter chain
-  //  matching rules <envoy_api_msg_listener.FilterChainMatch>`. For very simple configurations,
-  //  this field can still be used to select the cluster when no other matching rules are required.
-  //  Otherwise, a :ref:`deprecated_v1
-  //  <envoy_api_field_config.filter.network.tcp_proxy.v2.TcpProxy.deprecated_v1>` configuration is
-  //  required to use more complex routing in the interim.
-  //
-  string cluster = 2;
+  // [#not-implemented-hide:] Weighted clusters allow for specification of
+  // multiple upstream clusters along with weights that indicate the
+  // percentage of traffic to be forwarded to each cluster. TCP Proxy
+  // selects an upstream cluster based on the weights.
+  message WeightedCluster {
+    message ClusterWeight {
+      // Name of the upstream cluster. The cluster must exist in the
+      // :ref:`cluster manager configuration <config_cluster_manager>`.
+      string name = 1 [(validate.rules).string.min_bytes = 1];
+      
+      // An integer between 0 and :ref:`total_weight
+      // <envoy_api_field_route.WeightedCluster.total_weight>`. The choice of an upstream
+      // cluster is determined by its weight. The sum of weights across all entries in the
+      // clusters array must add up to the total_weight, which defaults to 100.
+      google.protobuf.UInt32Value weight = 2;
+
+      // [#not-implemented-hide:] Optional endpoint metadata match
+      // criteria. Only endpoints in the upstream cluster with metadata
+      // matching that set in metadata_match will be considered. The filter
+      // name should be specified as *envoy.lb*.
+      envoy.api.v2.core.Metadata metadata_match = 3;
+
+      // [#not-implemented-hide:] The per_filter_config field can be used
+      // to provide weighted cluster-specific configurations for
+      // filters. The key should match the filter name, such as
+      // *envoy.buffer* for the HTTP buffer filter. Use of this field is
+      // filter specific; see the :ref:`HTTP filter documentation
+      // <config_http_filters>` for if and how it is utilized.
+      map<string, google.protobuf.Struct> per_filter_config = 4;
+    }
+
+    // Specifies one or more upstream clusters associated with the TCP proxy.
+    repeated ClusterWeight clusters = 1 [(validate.rules).repeated .min_items = 1];
+
+    // Specifies the total weight across all clusters. The sum of all cluster
+    // weights must equal this value, which must be greater than 0. Defaults
+    // to 100.
+    google.protobuf.UInt32Value total_weight = 3 [(validate.rules).uint32.gte = 1];
```

Comment:
Do we need `total_weight`, or can it be computed from the `clusters` definition?

Reply:
I think this is to allow people to blackhole a certain portion of traffic. I copied this verbatim from http so that we can maintain the same semantics for http thrift and tcp

Reply:
Got it.

Just FYI, it's determined from the `clusters` definition for thrift currently: https://github.com/envoyproxy/envoy/blob/c32aed93f5d27e8cbd17c1c47cf928d7e8115af4/api/envoy/config/filter/network/thrift_proxy/v2alpha1/route.proto#L120.

Reply:
Oh bummer. Why have different weighted clusters! I will leave it upto you.

Reply:
But this is documented as "the sum of all cluster weights must equal this value". So there's no use for blackholing then, right? It seems like for blackholing, you can create a cluster with no hosts (and even name it "blackhole"), and give it a specific weight.

Reply:
That feels like a better approach to me. Thanks.